File: Chapitre2/exercice2.5.py
import os
import sys
import time
import random

def main():

    # ① Durées aléatoires entre 1 et 10 secondes
    duree1 = random.randint(1, 10)
    duree2 = random.randint(1, 10)

    print("=== Démarrage ===")
    print(f"Fils 1 va dormir {duree1} secondes")
    print(f"Fils 2 va dormir {duree2} secondes")
    print(f"Le plus long est le fils "
          f"{'1' if duree1 > duree2 else '2'} "
          f"({max(duree1, duree2)} secondes)\n")

    # ② Mémoriser le temps de départ
    debut = time.time()

    # ③ Créer fils 1
    pid1 = os.fork()
    if pid1 == -1:
        print("Erreur fork fils 1")
        sys.exit(1)

    if pid1 == 0:
        # ── FILS 1 ──────
        print(f"[Fils 1] démarre (PID = {os.getpid()})")
        time.sleep(duree1)
        print(f"[Fils 1] terminé après {duree1} secondes")
        sys.exit(0)

    # ④ Créer fils 2
    pid2 = os.fork()
    if pid2 == -1:
        print("Erreur fork fils 2")
        sys.exit(1)

    if pid2 == 0:
        # ── FILS 2 ─
        print(f"[Fils 2] démarre (PID = {os.getpid()})")
        time.sleep(duree2)
        print(f"[Fils 2] terminé après {duree2} secondes")
        sys.exit(0)

    # ── PÈRE ──

    # ⑤ Attendre les deux fils
    for i in range(2):
        pid_termine, _ = os.wait()

        if pid_termine == pid1:
            print("[Père] fils 1 terminé")
        elif pid_termine == pid2:
            print("[Père] fils 2 terminé")

    # ⑥ Calculer la durée totale
    fin           = time.time()
    duree_totale  = fin - debut

    print("\n=== Résultat ===")
    print(f"Durée totale   : {duree_totale:.1f} secondes")
    print(f"Plus long fils : {max(duree1, duree2)} secondes")

if __name__ == "__main__":
    main()

File: Chapitre2/exercice2.5.c
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <sys/wait.h>
#include <sys/types.h>
#include <time.h>

int main() {

    srand(time(NULL));

    // ② Mémoriser le temps de départ
    time_t debut = time(NULL);

    // ③ Durées aléatoires entre 1 et 10 secondes
    int duree1 = 1 + rand() % 10;
    int duree2 = 1 + rand() % 10;

    printf("=== Démarrage ===\n");
    printf("Fils 1 va dormir %d secondes\n", duree1);
    printf("Fils 2 va dormir %d secondes\n", duree2);
    printf("Le plus long est le fils %d (%d secondes)\n\n",
           duree1 > duree2 ? 1 : 2,
           duree1 > duree2 ? duree1 : duree2);

    // ④ Créer fils 1
    pid_t pid1 = fork();
    if (pid1 == -1) { perror("fork fils1"); exit(1); }

    if (pid1 == 0) {
        // ── FILS 1 ──
        printf("[Fils 1] démarre (PID = %d)\n", getpid());
        sleep(duree1);
        printf("[Fils 1] terminé après %d secondes\n", duree1);
        exit(0);
    }

    // ⑤ Créer fils 2
    pid_t pid2 = fork();
    if (pid2 == -1) { perror("fork fils2"); exit(1); }

    if (pid2 == 0) {
        // ── FILS 2 
        printf("[Fils 2] démarre (PID = %d)\n", getpid());
        sleep(duree2);
        printf("[Fils 2] terminé après %d secondes\n", duree2);
        exit(0);
    }


    // ⑥ Attendre les deux fils
    // wait() attend LE PREMIER fils qui se termine
    // on attend deux fois pour attendre LES DEUX fils
    for (int i = 0; i < 2; i++) {
        pid_t pid_termine = wait(NULL);

        if (pid_termine == pid1)
            printf("[Père] fils 1 terminé\n");
        else if (pid_termine == pid2)
            printf("[Père] fils 2 terminé\n");
    }

    // ⑦ Calculer la durée totale
    time_t fin          = time(NULL);
    double duree_totale = difftime(fin, debut);

    printf("\n=== Résultat ===\n");
    printf("Durée totale : %.0f secondes\n", duree_totale);
    printf("(le plus long des deux : %d secondes)\n",
           duree1 > duree2 ? duree1 : duree2);

    return 0;
}